File: src/fg_agent_id/serde.py
"""Minimal validation/serialization helpers for the wire dataclasses.

These give the frozen dataclasses a pydantic-compatible surface
(``model_dump`` / ``model_validate`` / ``model_copy``) without a pydantic
dependency, so embedding protocols can keep their existing call sites.
"""

from __future__ import annotations

from datetime import UTC, datetime
from typing import Any


def require_str(name: str, value: Any) -> str:
    if not isinstance(value, str):
        raise TypeError(f"{name} must be a string, got {type(value).__name__}")
    return value


def require_datetime(name: str, value: Any) -> datetime:
    if not isinstance(value, datetime):
        raise TypeError(f"{name} must be a datetime, got {type(value).__name__}")
    if value.tzinfo is None:
        raise ValueError(f"{name} must be timezone-aware (naive datetimes are ambiguous)")
    return value


def normalize_timestamp(name: str, value: Any) -> datetime:
    """Coerce to the precision the wire format carries: UTC, milliseconds.

    Signed payloads render timestamps with :func:`canonical_timestamp`, which
    truncates to milliseconds. Normalizing at construction keeps the in-memory
    value byte-identical to what was (or will be) signed, so a round-trip
    through JSON never changes a signature's meaning.
    """
    dt = require_datetime(name, value)
    dt = dt.astimezone(UTC)
    return dt.replace(microsecond=(dt.microsecond // 1000) * 1000)


def canonical_timestamp(value: datetime) -> str:
    """RFC 3339 UTC with exactly three fractional digits and a ``Z`` suffix.

    Pinned precisely because these strings land inside signed payloads:
    Python's ``isoformat()`` varies its output (offset spelling, presence and
    width of fractional seconds), so two implementations signing "the same"
    instant would produce different bytes. One spelling, always.
    """
    dt = value.astimezone(UTC)
    return f"{dt:%Y-%m-%dT%H:%M:%S}.{dt.microsecond // 1000:03d}Z"


def parse_datetime(name: str, value: Any) -> datetime:
    """Accept a datetime or an RFC 3339 / ISO-8601 string (``Z`` suffix ok)."""
    if isinstance(value, datetime):
        return normalize_timestamp(name, value)
    if isinstance(value, str):
        try:
            parsed = datetime.fromisoformat(value)
        except ValueError as exc:
            raise ValueError(f"{name} is not a valid ISO-8601 datetime: {value!r}") from exc
        if parsed.tzinfo is None:
            raise ValueError(f"{name} must carry a timezone offset: {value!r}")
        return normalize_timestamp(name, parsed)
    raise TypeError(f"{name} must be a datetime or ISO-8601 string")


def require_mapping(name: str, value: Any) -> dict:
    if not isinstance(value, dict):
        raise TypeError(f"{name} must be a mapping, got {type(value).__name__}")
    return value

File: js/src/serde.ts
/** Timestamp handling pinned to the wire format (§3.2 of the spec).
 *
 * Timestamps inside signed payloads are RFC 3339 UTC with exactly three
 * fractional digits and a `Z` suffix. JS `Date` is already millisecond
 * precision and `toISOString()` emits exactly this spelling, so the work here
 * is validation: reject timezone-naive input and normalize offsets to UTC.
 */

export function canonicalTimestamp(value: Date): string {
  if (Number.isNaN(value.getTime())) throw new Error("invalid Date");
  return value.toISOString();
}

const OFFSET_RE =
  /^(\d{4})-(\d{2})-(\d{2})[T ](\d{2}):(\d{2}):(\d{2})(\.\d+)?(Z|z|[+-]\d{2}:?\d{2})$/;

/**
 * Parse an RFC 3339 / ISO-8601 timestamp. Rejects timezone-naive input
 * (naive datetimes are ambiguous) and truncates — not rounds — fractional
 * seconds to milliseconds, matching the Python reference.
 */
export function parseTimestamp(name: string, value: string | Date): Date {
  if (value instanceof Date) {
    if (Number.isNaN(value.getTime())) throw new Error(`${name} is an invalid Date`);
    return value;
  }
  if (typeof value !== "string") {
    throw new TypeError(`${name} must be a Date or ISO-8601 string`);
  }
  const match = OFFSET_RE.exec(value);
  if (!match) {
    throw new Error(
      `${name} must be an ISO-8601 datetime carrying a timezone offset: ${JSON.stringify(value)}`
    );
  }
  const [, y, mo, d, h, mi, s, frac, offset] = match;
  // Truncate fractional seconds to milliseconds (pad to at least 3 digits).
  const ms = frac ? Number((frac.slice(1) + "000").slice(0, 3)) : 0;
  let epoch = Date.UTC(Number(y), Number(mo) - 1, Number(d), Number(h), Number(mi), Number(s), ms);
  if (offset !== "Z" && offset !== "z") {
    const sign = offset!.startsWith("-") ? -1 : 1;
    const [oh, om] = offset!.slice(1).replace(":", "").match(/^(\d{2})(\d{2})$/)!.slice(1);
    epoch -= sign * (Number(oh) * 60 + Number(om)) * 60_000;
  }
  const parsed = new Date(epoch);
  if (Number.isNaN(parsed.getTime())) throw new Error(`${name} is not a valid datetime: ${value}`);
  return parsed;
}

export function requireString(name: string, value: unknown): string {
  if (typeof value !== "string") {
    throw new TypeError(`${name} must be a string, got ${typeof value}`);
  }
  return value;
}

export function requireMapping(name: string, value: unknown): Record<string, unknown> {
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    throw new TypeError(`${name} must be a plain object`);
  }
  return value as Record<string, unknown>;
}
